sql: Find(Index|Column)ByName cleanup

vivekmenezes opened the pull request.

vivekmenezes committed:
sqlbase: simplify ProcessDefaultColumns

vivekmenezes committed:
sqlbase: replace WRITE_ONLY with DELETE_AND_WRITE_ONLY state

vivekmenezes committed:
sqlbase: simplified Find{Index|Column}ByName()

The functions now return an index/column when it is in the active
or in the mutation list. It also returns true
when the index/column is been dropped.

Made Find{Index|Column}ByNormalizedName() internal.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16367)
<!-- Reviewable:end -->

vivekmenezes commented:
@jordanlewis any comments on this PR? Thanks

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> except for the odd signatures you've introduced that return both an error and a value - that seems a bit like an antipattern! I'd prefer that we find a way to remove that oddity.

---

Reviewed 1 of 1 files at r1, 8 of 8 files at r2, 8 of 8 files at r3, 8 of 8 files at r4, 8 of 8 files at r5.
Review status: 10 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sqlbase/structured.go, line 1524 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16367#-KmOsoHsLeXVwaspU1-w:-KmOsoHsLeXVwaspU1-x:b-3bk06p) ([raw file](https://github.com/cockroachdb/cockroach/blob/04e687d3a20b75038ea7056f85d12b251a77e939/pkg/sql/sqlbase/structured.go#L1524)):*
> ```Go
> // FindIndexByName finds the index with the specified name in the active
> // list or the mutations list. An index being dropped is returned with
> // an error: ErrIndexBeingDropped.
> ```

This seems very confusing - I don't think I know of any other signatures that both return an error *and* a value and expect callers to deal with that situation properly.

There must be a better way! Or, do we really need to return the index object if its being dropped?

---

*[pkg/sql/sqlbase/structured.go, line 1434 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16367#-KmOte6Aoq0Ixpofgv-M:-KmOte6BRJilUz0cQuZI:b-xkut6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd50c488fff281ea591be3e0634b0d569728441/pkg/sql/sqlbase/structured.go#L1434)):*
> ```Go
> 			if parser.ReNormalizeName(c.Name) == normName {
> 				if m.Direction == DescriptorMutation_DROP {
> 					return *c, ErrColumnBeingDropped
> ```

Same as before - this is confusing! Can we avoid returning both an error and a value?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16367#-:-KmOth6FQTBJNO8AbGRX:b-uvl36t)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I've changed the functions to return a boolean when returning a dropped element and a legitimate error when they can't find the element. TYFR!

---

Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/sqlbase/structured.go, line 1524 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16367#-KmOsoHsLeXVwaspU1-w:-KmU8Q9MfHcKmtJ8tjQc:b-5rxpfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/04e687d3a20b75038ea7056f85d12b251a77e939/pkg/sql/sqlbase/structured.go#L1524)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This seems very confusing - I don't think I know of any other signatures that both return an error *and* a value and expect callers to deal with that situation properly.

There must be a better way! Or, do we really need to return the index object if its being dropped?
</blockquote></details>

I've changed it to return a boolean for when it's returning a dropped element.

---

*[pkg/sql/sqlbase/structured.go, line 1434 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16367#-KmOte6Aoq0Ixpofgv-M:-KmU8jeWOuq1sN0NguUY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd50c488fff281ea591be3e0634b0d569728441/pkg/sql/sqlbase/structured.go#L1434)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Same as before - this is confusing! Can we avoid returning both an error and a value?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16367#-:-KmU8YMvZzNTcx0cv-jj:bav8k7j)*
<!-- Sent from Reviewable.io -->
